src/orchard/eldoc.clj

```diff
@@ -3,13 +3,20 @@
   in editors."
   {:author "Bozhidar Batsov"})
 
+(def java-interop-special-forms #{"." "set!" "new"})
+
 (defn- extract-arglists
   [info]
+  (println "Name: " (:name info)) ;DEBUG
   (cond
     (:special-form info) (->> (:forms info)
                               ;; :forms contains a vector of sequences or symbols
-                              ;; which we have to convert the format employed by :arglists
-                              (map #(if (coll? %) (vec %) (vector %))))
+                              ;; which we have to convert the format employed by :arglists.
+                              ;; We drop the first element of the vector in most cases
+                              ;; because that's the form itself. However, if the form is a
+                              ;; Java interop special form, we want to keep the first
+                              ;; element.
+                              (map #(if (coll? %) (vec (if (java-interop-special-forms (:name info)) % (rest %))) (vector %))))
```

Comment:
`orchard.eldoc/eldoc` is invoked by `cider.nrepl.middleware.info/eldoc-reply`, which calls `cider.nrepl.middleware.info/info` on the input and hands the result to `orchard.eldoc/eldoc`. 

To retrieve the data about the special form, `cider.nrepl.middleware.info/info` calls `orchard.info/info`, which is a thin wrapper around `orchard.info/info*`. `info*`, in turn, calls `orchard.info/clj-meta`. `clj-meta` calls `orchard.meta/special-sym-meta`. `special-sym-meta` calls `orchard.meta/repl-special-meta`, which does a require-and-resolve trick to get at the private `clojure.repl/special-doc`, which is where the `info` map's data actually comes from. And, sure enough, if we do:

```clojure
(let [f (orchard.meta/require-and-resolve 'clojure.repl/special-doc)]
  (f 'if))
```

we get

```
{:forms [(if test then else?)],
 :doc
 "Evaluates test. If not the singular values nil or false,\n  evaluates and yields then, otherwise, evaluates and yields else. If\n  else is not supplied it defaults to nil.",
 :name if,
 :special-form true}
```

Note that the first element in the `forms` list is `if` itself.

I could change the behavior of `clojure.info/info*` to filter out the special form itself from the `:forms`, but that would be a more wide-ranging change, because `cider.nrepl.middleware.info/info` is also called by `cider.nrepl.middleware.info/info-reply`, which is the function that handles `"info"` requests. I don't know what relies on making `info` requests to the `cider.nrepl` middleware, so changing the data that `info` returns has the potential to break those users. Changing the behavior of `eldoc-reply`, seems to me a narrower change even though it is at the wrong layer, as you say.